File: Assignment1/main.py
import numpy as np
from scipy.stats import spearmanr as spear
import time

# Vocabulary paths
VOCAB_PATH = './cs402_assn1/vocab-3k.txt'
WORDSIM_PATH = './cs402_assn1/vocab-wordsim.txt'

# Corpus path
CORPUS_PATH = 'wiki-0.1percent.txt'

# Similarity dataset paths
MEN_PATH = './cs402_assn1/men.txt'
SIMLEX_PATH = './cs402_assn1/simlex-999.txt'


def load_vocab(wordsim=True):
    try:
        PATH = WORDSIM_PATH if wordsim else VOCAB_PATH
        vocab = np.loadtxt(PATH, dtype=str)
        return vocab
    except Exception as e:
        raise FileNotFoundError(e)


def load_simdataset(men=True):
    try:
        PATH = MEN_PATH if men else SIMLEX_PATH
        pairs = np.loadtxt(PATH, dtype=str, skiprows=1, usecols=(0, 1))
        corr = np.loadtxt(PATH, dtype=float, skiprows=1, usecols=(2))
        return pairs, corr
    except Exception as e:
        raise FileNotFoundError(e)


def cosine_similarity(C, V, men=True):
    pairs, corr = load_simdataset(men)
    M = np.zeros_like(corr, dtype=float)
    sqr_len = C.shape[1]**2
    for i, pair in enumerate(pairs):
        if pair[0] in V and pair[1] in V:
            idx_1 = np.where(V == pair[0])[0][0]
            idx_2 = np.where(V == pair[1])[0][0]
            M[i] = np.dot(C[idx_1], C[idx_2])/sqr_len
    return M


def EVALWS(M):
    VX = load_vocab(True)
    VY = load_vocab(False)
    men = cosine_similarity(M, VX, True)
    simlex = cosine_similarity(M, VX, False)
    _, men_corr = load_simdataset(True)
    _, simlex_corr = load_simdataset(False)
    print("Correlation for MEN")
    print(spear(men_corr, men))
    print("Correlation for SIMLEX")
    print(spear(simlex_corr, simlex))


def nearest_neighbors(C, word, num=10):
    V = load_vocab(False)
    check = np.where(V == word)[0]
    if check.shape[0] == 0:
        raise AssertionError("No such word in vocabulary")
    idx = check[0]
    length = V.shape[0]
    sqr_len = length**2
    L = np.empty_like(V, dtype=float)
    for i in range(length):
        L[i] = np.dot(C[idx], C[i])/sqr_len
    D = dict()
    max_idx = L.argmax(0)
    L[max_idx] = -np.inf
    for i in range(num):
        max_idx = L.argmax(0)
        D[V[max_idx]] = L[max_idx]
        L[max_idx] = -np.inf
    return D


def print_dict(D):
    for key in D:
        tab = '\t' if len(key) > 8 else '\t\t'
        print(f"{str(key)}{tab}: {float(D[key])}")


class CorpusDataset:
    def __init__(self, chunk_size):
        self.corpus_size = 0
        self.chunk_size = chunk_size
        self.skip = 0
        self.empty = False
        self.switch = True
        self.left = None
        self.right = self._load_data()

    def is_empty(self):
        return self.empty

    def size(self):
        return self.corpus_size

    def load_chunk(self):
        if self.is_empty():
            raise RuntimeError("No chunks to load")
        else:
            if self.switch:
                self.left = self._load_data()
                self.switch = False
                return self.right
            else:
                self.right = self._load_data()
                self.switch = True
                return self.left

    def _load_data(self):
        try:
            corpus = np.genfromtxt(
                CORPUS_PATH, dtype=str, delimiter='\n', skip_header=self.skip, max_rows=self.chunk_size, encoding='utf8')
            if corpus.shape[0] == 0:
                self.empty = True
            self.skip += self.chunk_size
            self.corpus_size += corpus.shape[0]
            return corpus
        except Exception as e:
            raise FileNotFoundError(e)


class Counter:
    def __init__(self, window=3, chunk_size=100000, flag=True):
        self.START = 0
        self.END = 1
        self.WINDOW = window
        self.CORPUS_SIZE = 0
        self.CHUNK_SIZE = chunk_size
        self.CORPUS = CorpusDataset(self.CHUNK_SIZE)
        self.VX = load_vocab(flag)
        self.VY = load_vocab(False)
        self.C = np.zeros((self.VX.shape[0], self.VY.shape[0]), dtype=float)
        self.C_PMI = None
        self.VX_COUNT = np.zeros_like(self.VX, dtype=int)
        self.VY_COUNT = np.zeros_like(self.VY, dtype=int)
        self.check_pmi = False

    def count_raw(self):
        self.check_pmi = True
        chunk_num = 0
        while not self.CORPUS.is_empty():
            chunk = self.CORPUS.load_chunk()
            chunk_num += 1
            print(f"Chunk {chunk_num} is loaded")
            x = 0
            for sen in chunk:
                sentence = sen.split(' ')
                """for pos, word in enumerate(sentence):
                    found = np.where(self.VX == word)[0]
                    if found.size > 0:
                        self.VX_COUNT[found[0]] += 1
                        row = self.C[found[0]]
                        for i in range(self.WINDOW):
                            c_pos = pos + i + 1
                            if c_pos >= length:
                                row[self.END] += 1
                                break
                            c_idx = np.where(self.VY == sentence[c_pos])
                            if len(c_idx) != 0:
                                prev.append(c_idx[0])
                                self.VY_COUNT[c_idx[0]] += 1
                                row[c_idx[0]] += 1
                        for i in range(self.WINDOW + 2, 2 * self.WINDOW + 2):
                            if i > len(prev):
                                row[self.START] += 1
                                break
                            self.VY_COUNT[prev[-i]] += 1
                            row[prev[-i]] += 1"""
                """XI = []
                YI = []
                for word in sentence:
                    xi = np.where(self.VX == word)[0]
                    yi = np.where(self.VY == word)[0]
                    if xi.size > 0:
                        XI.append(xi[0])
                    if yi.size > 0:
                        YI.append(yi[0])"""

                if (x % 10000 == 0):
                    size = chunk.shape[0]
                    print(f"Chunk {chunk_num} progress: {x/size*100:0.0f}%")
                x += 1
        self.CORPUS_SIZE = self.CORPUS.size()
        return self.C

    def calculate_pmi(self):
        if self.check_pmi:
            VX_PROB = np.expand_dims(self.VX_COUNT/self.CORPUS_SIZE, 1)
            VY_PROB = np.expand_dims(self.VY_COUNT/self.CORPUS_SIZE, 0)
            VX_VY_PROB = np.matmul(VX_PROB, VY_PROB)
            VX_VY_PROB[VX_VY_PROB == 0] = 1
            C_PROB = self.C/self.CORPUS_SIZE
            self.C_PMI = np.log2(np.divide(C_PROB, VX_VY_PROB))
            self.C_PMI[self.C_PMI < 0] = 0
            return self.C_PMI
        else:
            raise AssertionError("Counting is not performed")


# 1.1 Distributional Counting (takes around 30 minutes to run)
counter_3 = Counter(3)
tic = time.perf_counter()
C_3 = counter_3.count_raw()
toc = time.perf_counter()
print(f"Timer: {toc - tic:0.0f} seconds")
EVALWS(C_3)

# 1.2 Computing PMIs (runs relativelly fast)
C_PMI_3 = counter_3.calculate_pmi()
EVALWS(C_PMI_3)
del counter_3

"""# 1.3 Experimentation (takes around 70 minutes to run)
counter_1 = Counter(1)
C_1 = counter_1.count_raw()
EVALWS(C_1)
C_PMI_1 = counter_1.calculate_pmi()
EVALWS(C_PMI_1)
del counter_1

counter_6 = Counter(6)
C_6 = counter_6.count_raw()
EVALWS(C_6)
C_PMI_6 = counter_6.calculate_pmi()
EVALWS(C_PMI_6)
del counter_6

# 1.4.1 Warm-up: Printing nearest neighbors (takes around a week to run)

counter_1 = Counter(1, flag=False)
CN_1 = counter_1.count_raw()
CN_PMI_1 = counter_1.calculate_pmi()
D_1 = nearest_neighbors(CN_PMI_1, 'monster', 10)
print_dict(D_1)
del counter_1

counter_6 = Counter(6, flag=False)
CN_6 = counter_6.count_raw()
CN_PMI_6 = counter_6.calculate_pmi()
D_6 = nearest_neighbors(CN_PMI_6, 'monster', 10)
print_dict(D_6)
del counter_6
"""

File: Assignment1/main.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <cmath>
#include <algorithm>
#include <chrono>

typedef double Data;
const std::string VOCAB_PATH = "./cs402_assn1/vocab-25k.txt";
const std::string WORDSIM_PATH = "./cs402_assn1/vocab-wordsim.txt";
const std::string CORPUS_PATH = "wiki-1percent.txt";
const std::string MEN_PATH = "./cs402_assn1/men.txt";
const std::string SIMLEX_PATH = "./cs402_assn1/simlex-999.txt";

void load_vocab(std::vector<std::string> *vocab, bool wordsim = true)
{
    std::string PATH = wordsim ? WORDSIM_PATH : VOCAB_PATH;
    std::ifstream vocab_file(PATH);
    if (!vocab_file.is_open())
    {
        std::cout << PATH << " not found" << std::endl;
        exit(1);
    }
    std::string buffer;
    while (getline(vocab_file, buffer))
    {
        vocab->push_back(buffer);
    }
    vocab_file.close();
    std::sort(vocab->begin(), vocab->end());
    return;
}

std::vector<std::string> split(const std::string &buffer)
{
    std::vector<std::string> buffer_split;
    buffer_split.push_back("");
    for (int i = 0; i < buffer.length(); i++)
    {
        if (buffer[i] == '\t' || buffer[i] == ' ')
        {
            buffer_split.push_back("");
            continue;
        }
        buffer_split.back().push_back(buffer[i]);
    }
    return buffer_split;
}

int find_index(const std::string &word, const std::vector<std::string> &vocab)
{
    int low = 0;
    int high = vocab.size();
    int mid = (low + high) / 2;
    while (low + 1 < high)
    {
        if (word == vocab[mid])
        {
            return mid;
        }
        else if (word < vocab[mid])
        {
            high = mid;
            mid = (low + high) / 2;
        }
        else
        {
            low = mid;
            mid = (low + high) / 2;
        }
    }
    return -1;
}

void load_simdataset(std::vector<std::vector<std::string>> *pairs, std::vector<Data> *corrs, bool men = true, bool skip = true)
{
    std::string PATH = men ? MEN_PATH : SIMLEX_PATH;
    std::ifstream sim_file(PATH);

    if (!sim_file.is_open())
    {
        std::cout << PATH << " not found" << std::endl;
        exit(1);
    }
    std::string buffer;
    std::vector<std::string> buffer_split;
    std::vector<std::string> pair;
    if (skip)
        std::getline(sim_file, buffer);
    while (std::getline(sim_file, buffer))
    {

        buffer_split = split(buffer);
        pair.push_back(buffer_split[0]);
        pair.push_back(buffer_split[1]);

        pairs->push_back(pair);
        corrs->push_back(std::stod(buffer_split[2]));
    }
    sim_file.close();
    return;
}

std::vector<Data> cosine_similarity(const std::vector<std::vector<Data>> &C, const std::vector<std::string> &V, bool men = true)
{
    std::vector<std::vector<std::string>> pairs;
    std::vector<Data> corrs;
    load_simdataset(&pairs, &corrs);
    std::vector<Data> my_coors;
    Data sqr_len = static_cast<Data>(std::sqrt(C[0].size()));
    int idx_1;
    int idx_2;
    for (int i = 0; i < pairs.size(); i++)
    {
        idx_1 = find_index(pairs[i][0], V);
        idx_2 = find_index(pairs[i][1], V);
        if (idx_1 > 0 && idx_2 > 0)
        {
            for (int x = 0; x < pairs.size(); x++)
            {
                my_coors.push_back(C[idx_1][x] * C[idx_2][x] / sqr_len);
            }
        }
    }
    return my_coors;
}

class Counter
{
private:
    const int START = 0;
    const int END = 1;
    int WINDOW;
    Data CORPUS;
    std::vector<std::string> VX;
    std::vector<std::string> VY;
    std::vector<std::vector<Data>> C;
    std::vector<std::vector<Data>> C_PMI;
    std::vector<Data> VX_COUNT;
    std::vector<Data> VY_COUNT;
    bool check_pmi = false;
    int num_rows;
    int num_cols;

public:
    Counter(int window, bool flag = true) : WINDOW(window)
    {
        load_vocab(&VX, flag);
        load_vocab(&VY, false);
        num_rows = VX.size();
        num_cols = VY.size();
        C.resize(num_rows);
        for (auto &row : C)
            row.resize(num_cols);
        C_PMI.resize(num_rows);
        for (auto &row : C_PMI)
            row.resize(num_cols);
        VX_COUNT.resize(num_rows);
        VY_COUNT.resize(num_cols);
    }

    std::vector<std::vector<Data>> *count_raw()
    {
        check_pmi = true;
        std::ifstream corpus_file(CORPUS_PATH);
        if (!corpus_file.is_open())
        {
            std::cout << CORPUS_PATH << " not found" << std::endl;
            exit(1);
        }
        std::string buffer;
        int count = 0;
        while (getline(corpus_file, buffer))
        {
            CORPUS++;
            std::vector<std::string> sentence = split(buffer);
            int pos = 0;
            int size = sentence.size();
            for (std::string &word : sentence)
            {
                int found = find_index(word, VX);
                if (found > 0)
                {
                    VX_COUNT[found]++;
                    std::vector<Data> &row = C[found];
                    for (int i = (pos <= WINDOW ? 0 : pos - WINDOW); i < (size <= pos + WINDOW ? size : pos + WINDOW); i++)
                    {
                        if (i == pos)
                        {
                            continue;
                        }
                        int next = find_index(sentence[i], VY);
                        if (next > 0)
                        {
                            VY_COUNT[next]++;
                            row[next]++;
                        }
                    }
                }
                pos++;
            }

            count++;
            if (count % 100000 == 0)
                std::cout << "Progress bar: " << count / 10000 << "%" << std::endl;
        }
        corpus_file.close();
        return &C;
    }

    std::vector<std::vector<Data>> *calculate_pmi()
    {
        Data temp;
        if (check_pmi)
        {
            for (int x = 0; x < num_rows; x++)
            {
                for (int y = 0; y < num_cols; y++)
                {
                    temp = (VX_COUNT[x] * VY_COUNT[y]) / CORPUS;
                    if (temp == 0)
                        temp = 1;
                    C_PMI[x][y] = std::log2(C[x][y] / temp);
                }
            }
            return &C_PMI;
        }
        else
        {
            std::cout << "Counting is performed\n";
            return nullptr;
        }
    }
};

int main()
{
    Counter counter(1, false);
    std::vector<std::vector<Data>> *C = counter.count_raw();
    std::vector<std::vector<Data>> *C_PMI = counter.calculate_pmi();

    return 0;
}
